File: py/math/PollardRho.py
#!/usr/bin/env python
# -*- coding:gbk -*-

import random
from typing import List

class PollardRho:
    def __init__(self):
        self.__primes = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29]

    def solution(self, num: int) -> List[int]:
        result = []
        if num > 1:
            self.__factorization(num, 256, result)
        result.sort()
        return result

    def __factorization(self, num: int, aux: int, result: List[int]) -> None:
        if self.__isPrime(num):
            result.append(num)
            return

        factor = num
        while factor >= num:
            factor = self.__getFactor(factor, aux - 1)
            aux -= 1
        self.__factorization(factor, aux, result)
        self.__factorization(num // factor, aux, result)

    def __isPrime(self, num: int) -> bool:
        if num in self.__primes:
            return True
        if num < 2 or not num & 1:
            return False

        base, index = num - 1, 0
        while not base & 1:
            base >>= 1
            index += 1

        for value in self.__primes:
            if self.__check(value, num, base, index):
                return False
        return True

    def __getFactor(self, num: int, aux: int) -> int:
        pivot = random.randrange(1, num - 1)
        value, index, base = (self.__multiplyMod(pivot, pivot, num) + aux) % num, 1, 2
        while value != pivot:
            factor = self.__gcd(pivot - value, num)
            if factor > 1 and factor < num:
                return factor

            index += 1
            if index == base:
                pivot = value
                base <<= 1
            value = (self.__multiplyMod(value, value, num) + aux) % num
        return num

    def __check(self, factor: int, num: int, base: int, index: int) -> bool:
        pivot = self.__powerMod(factor, base, num)
        if 1 == pivot:
            return False

        while index:
            value = self.__multiplyMod(pivot, pivot, num)
            if 1 == value and 1 != pivot and pivot != num - 1:
                return True
            pivot = value
            index -= 1
        return 1 != pivot

    def __powerMod(self, base: int, index: int, mod: int) -> int:
        base %= mod
        if 1 == index:
            return base

        aux, result = base, 1
        while index:
            if index & 1:
                result = self.__multiplyMod(result, aux, mod)
            aux = self.__multiplyMod(aux, aux, mod)
            index >>= 1
        return result

    def __multiplyMod(self, i: int, j: int, mod: int) -> int:
        i %= mod
        j %= mod
        result = 0
        while j:
            if j & 1:
                result = (result + i) % mod
            i <<= 1
            if i >= mod:
                i %= mod
            j >>= 1
        return result

    def __gcd(self, num1: int, num2: int) -> int:
        return num1 if not num2 else self.__gcd(num2, num1 % num2)

File: java/math/PollardRho.java
package math;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Random;

public class PollardRho {
    private static final int[] _primes = {2, 3, 5, 7, 11, 13, 17, 19, 23, 29};
    private static final Random _random = new Random();

    public long[] solution(long num) {
        List<Long> result = new ArrayList<>();
        if (num > 1) {
            factorization(num, 256, result);
        }
        Collections.sort(result);
        return result.stream().mapToLong(Long::valueOf).toArray();
    }

    private void factorization(long num, int aux, List<Long> result) {
        if (isPrime(num)) {
            result.add(num);
            return;
        }

        long factor = num;
        while (factor >= num) {
            factor = getFactor(factor, aux - 1);
            --aux;
        }
        factorization(factor, aux, result);
        factorization(num / factor, aux, result);
    }

    private boolean isPrime(long num) {
        if (exists(num)) {
            return true;
        }
        if (num < 2 || 0 == (num & 1)) {
            return false;
        }

        long base = num - 1;
        int index = 0;
        while (0 == (base & 1)) {
            base >>= 1;
            ++index;
        }

        for (int value : _primes) {
            if (check((long) value, num, base, index)) {
                return false;
            }
        }
        return true;
    }

    private long getFactor(long num, int aux) {
        long pivot = Math.abs(_random.nextLong()) % (num - 1) + 1;
        long value = (multiplyMod(pivot, pivot, num) + aux) % num;
        long index = 1;
        long base = 2;
        while (value != pivot) {
            long factor = gcd(pivot - value, num);
            if (factor > 1 && factor < num) {
                return factor;
            }

            ++index;
            if (index == base) {
                pivot = value;
                base <<= 1;
            }
            value = (multiplyMod(value, value, num) + aux) % num;
        }
        return num;
    }

    private boolean check(long factor, long num, long base, int index) {
        long pivot = powerMod(factor, base, num);
        if (1 == pivot) {
            return false;
        }

        while (0 != index) {
            long value = multiplyMod(pivot, pivot, num);
            if (1 == value && 1 != pivot && pivot != num - 1) {
                return true;
            }
            pivot = value;
            --index;
        }
        return 1 != pivot;
    }

    private long powerMod(long base, long index, long mod) {
        base %= mod;
        if (1 == index) {
            return base;
        }

        long aux = base;
        long result = 1;
        while (0 != index) {
            if (0 != (index & 1)) {
                result = multiplyMod(result, aux, mod);
            }
            aux = multiplyMod(aux, aux, mod);
            index >>= 1;
        }
        return result;
    }

    private long multiplyMod(long i, long j, long mod) {
        i %= mod;
        j %= mod;
        long result = 0;
        while (0 != j) {
            if (0 != (j & 1)) {
                result = (result + i) % mod;
            }
            i <<= 1;
            if (i >= mod) {
                i %= mod;
            }
            j >>= 1;
        }
        return result;
    }

    private long gcd(long num1, long num2) {
        return (0 ==num2) ? num1 : gcd(num2, num1 % num2);
    }

    private boolean exists(long num) {
        int start = 0;
        int end = _primes.length - 1;
        while (start <= end) {
            int middle = start + ((end - start) >> 1);
            if (num == _primes[middle]) {
                return true;
            }

            if (num < _primes[middle]) {
                end = middle - 1;
            } else {
                start = middle + 1;
            }
        }
        return false;
    }
}
